internal/core/app_status_report.go

```diff
@@ -0,0 +1,340 @@
+		c = nil
+		err = nil
+	}
+	if err != nil {
+		a.logger.Error("error creating component in platform", "error", err)
+		return nil, nil, err
+	}
+
+	a.logger.Debug("starting status report operation")
+	statusReporter, ok := c.Value.(component.Status)
+
+	if !ok || statusReporter.StatusFunc() == nil {
+		a.logger.Debug("component is not a Status or has no StatusFunc()")
+		return nil, nil, nil
+	}
+	defer c.Close()
+
+	return a.statusReport(ctx, "release_statusreport", c, releaseTarget)
+}
+
+// A generic status report func that takes an already setup component and a
+// specific target to execute the report on.
+func (a *App) statusReport(
+	ctx context.Context,
+	loggerName string,
+	component *Component,
+	target interface{},
+) (*pb.StatusReport, *sdk.StatusReport, error) {
+	if loggerName == "" {
+		loggerName = "statusreport"
+	}
+
+	result, msg, err := a.doOperation(ctx, a.logger.Named(loggerName), &statusReportOperation{
+		Component: component,
+		Target:    target,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var status *sdk.StatusReport
+	if result != nil {
+		status = result.(*sdk.StatusReport)
+	}
+
+	return msg.(*pb.StatusReport), status, nil
+}
+
+// Sets up the eval context for a status report for deployments
+func (a *App) deployStatusReportEvalContext(
+	ctx context.Context,
+	d *pb.Deployment,
+	evalCtx *hcl.EvalContext,
+) error {
+	// Query the artifact
+	var artifact *pb.PushedArtifact
+	if d != nil && d.Preload != nil {
+		artifact = d.Preload.Artifact
+	}
+	if artifact == nil && d != nil {
+		a.logger.Debug("querying artifact", "artifact_id", d.ArtifactId)
+		resp, err := a.client.GetPushedArtifact(ctx, &pb.GetPushedArtifactRequest{
+			Ref: &pb.Ref_Operation{
+				Target: &pb.Ref_Operation_Id{
+					Id: d.ArtifactId,
+				},
+			},
+		})
+		if err != nil {
+			a.logger.Error("error querying artifact",
+				"artifact_id", d.ArtifactId,
+				"error", err)
+			return err
+		}
+
+		artifact = resp
+	}
+
+	evalCtx.Variables = map[string]cty.Value{}
+	if err := evalCtxTemplateProto(evalCtx, "artifact", artifact); err != nil {
+		a.logger.Warn("failed to prepare template variables, will not be available",
+			"err", err)
+	}
+	return nil
+}
+
+// Sets up the eval context for a status report for deployments
+func (a *App) releaseStatusReportEvalContext(
+	ctx context.Context,
+	r *pb.Release,
+	evalCtx *hcl.EvalContext,
+) error {
+	// Query the deployment
+	a.logger.Debug("querying deployment", "deployment_id", r.DeploymentId)
+	resp, err := a.client.GetDeployment(ctx, &pb.GetDeploymentRequest{
+		Ref: &pb.Ref_Operation{
+			Target: &pb.Ref_Operation_Id{
+				Id: r.DeploymentId,
+			},
+		},
+
+		LoadDetails: pb.Deployment_ARTIFACT,
+	})
+	if status.Code(err) == codes.NotFound {
+		// TODO: fix panic. Should this be an error though?
+		resp = nil
+		err = nil
+		a.logger.Warn("deployment not found, will attempt status report regardless",
+			"deployment_id", r.DeploymentId)
```

Comment:
Best solution to the deref crash of `resp` below is probably just return and not set the variables here.